### PR TITLE
sae: Implement `Tx.TransferNonAVAX`

### PR DESCRIPTION
**Pull request by StephenButtolph**
## Why this should be merged

This PR factors out the non-AVAX ANT asset transfers from #5303.

## How this works

In coreth, all state changes lived under a single `EVMStateTransfer` function. In SAE, we need to split this up into:
1. Ethereum state changes (nonce + AVAX transfers)
2. Non-AVAX transfers.

`AsOp` already introduced (1), so this PR adds (2). For compatibility, this PR fuzzes against `EVMStateTransfer` and `AsOp` + `TransferNonAVAX`.

## How this was tested

- [X] Extended tx test vector
- [X] Fuzzed against coreth

## Need to be documented in RELEASES.md?

No

### CODE REVIEW

**Review by powerslider — APPROVED** [review]
LGTM! Non-AVAX transfers now have their own dedicated method, with a fuzz test diffing the full SAE flow against coreth's reference. Nothing else to suggest as improvements.

File: vms/saevm/cchain/tx/export.go [hunk 1]
```diff
@@ -6,12 +6,14 @@ package tx
...
 	"errors"
 	"fmt"
+	"math/big"
 
 	"github.com/ava-labs/libevm/common"
 
 	// Imported for [atomic.UnsignedExportTx.Burned] comment resolution.
 	_ "github.com/ava-labs/avalanchego/graft/coreth/plugin/evm/atomic"
 
+	"github.com/ava-labs/avalanchego/graft/coreth/core/extstate"
```
Comment (alarso16) [inline vms/saevm/cchain/tx/export.go:16]: I thought we weren't import graft?

File: vms/saevm/cchain/tx/tx_test.go [hunk 2]
```diff
@@ -31,6 +38,11 @@ import (
 	safemath "github.com/ava-labs/avalanchego/utils/math"
 )
 
+func TestMain(m *testing.M) {
+	customtypes.Register()
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx_test.go]: Why are registering this? I think what you meant to register is `extstate`

File: vms/saevm/cchain/tx/tx.go [hunk 3]
```diff
@@ -36,9 +37,13 @@ type Tx struct {
...
 // Unsigned is a common interface implemented by [Import] and [Export].
 //
-// TODO(StephenButtolph): Expand this interface to include UTXO handling,
-// verification, and state execution.
+// TODO(StephenButtolph): Expand this interface to include UTXO handling and
+// verification.
 type Unsigned interface {
+	// TransferNonAVAX transfers the non-AVAX balances requested by this
+	// transaction.
+	TransferNonAVAX(avaxAssetID ids.ID, statedb *extstate.StateDB) error
```
Comment (alarso16) [inline vms/saevm/cchain/tx/tx.go:47]: Why is this exported?

File: vms/saevm/cchain/tx/tx.go
Comment (alarso16) [inline vms/saevm/cchain/tx/tx.go:59]: Seems weird we aren't including these as an `op`. Why? Is it because of the multicoin-db weirdness?

File: vms/saevm/cchain/tx/tx_test.go (hunk 2 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx_test.go]: It's needed for the multicoin stuff.

File: vms/saevm/cchain/tx/export.go (hunk 1 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:16]: Because of the transition logic, we need to share the libevm registrations during the SAE migration.

We can either:
1. Have SAE depend on coreth for the libevm registrations
2. Move the libevm registrations into SAE and have coreth depend on SAE.

I felt like (1) was less work and wouldn't require moving anything non-SAE related into the SAE codebase (since I presume there is some sprawl in these areas that will be removed once we are in an SAE only world).

But you're right that we should be extremely conservative with any dependencies on coreth. I think it should only be for the libevm registrations.

File: vms/saevm/cchain/tx/tx.go (hunk 3 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:47]: This is exported because it is consumed by external users (specifically, it will be called in the `AfterBlock` hook during block execution.

The other functions _aren't_ exported because they are consumed only to support other (exported) functions on the Tx type. `AtomicRequests` was the closest to also being exported on this type, but because it consumed the `Tx.ID`, it didn't make sense for consumers to call multiple functions (imo).

Technically, we could make this unexported on the unsigned txs, so that we only export it on the signed tx... But we actually don't need a signed tx for calling this function... So it felt a bit weird to me to do that...

Although perhaps it would make sense to just reduce the number of exported identifiers (currently there are 3, we could have just 1).

File: vms/saevm/cchain/tx/tx.go
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:59]: I think the comment on `asOp` is definitely wrong in this PR... So we should certainly clean it up at the very least.

We could change the way this function is implemented so that we just return a couple maps of changes (and then we wouldn't need this part of the code to ever have access to EVM state)... But we'd then just essentially need to take the current implementation of these functions and implement them the same way (but just over a map).

I'm not sure that that is better...

File: vms/saevm/cchain/tx/tx.go (hunk 3 above)
Comment (alarso16) [reply vms/saevm/cchain/tx/tx.go:47]: That seems reasonable. It would be weird to put it on the signed tx, so that doesn't seem right. Being included in the AfterBlock hooks is sufficient justification to me

File: vms/saevm/cchain/tx/tx.go
Comment (alarso16) [reply vms/saevm/cchain/tx/tx.go:59]: Hmm yeah I'm not sure it's better either. It doesn't seem like a logical thing to include in a `hook.Op`, so we can leave it.

File: vms/saevm/cchain/tx/tx_test.go (hunk 2 above)
Comment (alarso16) [reply vms/saevm/cchain/tx/tx_test.go]: I still think we need to register the `extstate` extras. I don't think this is an accurate reflection of using the statedb if we don't

File: vms/saevm/cchain/tx/export.go (hunk 1 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/export.go:16]: This seems like a slippery slope and I don't like importing graft -- I feel like we could get lazy with this. Can we have an SAE package that will be explicitly deprecated and must be removed for things to be 'good', in a similar way we did the RPC stubs and temporary package.

Comment (powerslider) [reply vms/saevm/cchain/tx/export.go:16]: The options here are either to just start working on the migration of those packages or just to decouple with port interfaces. We will inevitably hit these issues so we just need to decide if we will lean on dependency inversion (at least temporary) or just accept importing from graft.
Here is an example of how I've applied this pattern in `graft/evm` when I did not want to import any `graft/coreth` packages in the code.

- Port interface in `graft/evm`: https://github.com/ava-labs/avalanchego/blob/master/graft/evm/sync/engine/client.go#L69
- Adapter part in `graft/coreth`: https://github.com/ava-labs/avalanchego/blob/master/graft/coreth/plugin/evm/sync_adapter.go

This way my only dependency is the interface while the implementation sits in `graft/coreth`.

File: vms/saevm/cchain/tx/compatibility_test.go [hunk 4]
```diff
@@ -15,19 +16,29 @@ import (
...
 	"github.com/ava-labs/avalanchego/vms/components/gas"
 	"github.com/ava-labs/avalanchego/vms/saevm/cchain/tx/txtest"
+	"github.com/ava-labs/avalanchego/vms/saevm/cmputils"
 	"github.com/ava-labs/avalanchego/vms/saevm/hook"
 
 	. "github.com/ava-labs/avalanchego/vms/saevm/cchain/tx"
 )
 
-// fuzz seeds f with [NewTxs] and fuzzes the test.
+// fuzz seeds f with [NewTxs], specifies simple alphabets used to bias the
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/compatibility_test.go:30]:
```suggestion
// fuzz seeds f with `NewTxs`, specifies simple alphabets used to bias the
```

I do this all the time, it's so much easier to be critical when you review rather than write for some reason

File: vms/saevm/cchain/tx/export.go [hunk 5]
```diff
@@ -136,3 +138,22 @@ func (e *Export) atomicRequests(txID ids.ID) (ids.ID, *chainsatomic.Requests, er
 	}
 	return e.DestinationChain, &chainsatomic.Requests{PutRequests: elems}, nil
 }
+
+var errInsufficientFunds = errors.New("insufficient funds")
+
+// TransferNonAVAX subtracts the non-AVAX balances from the statedb.
+func (e *Export) TransferNonAVAX(avaxAssetID ids.ID, statedb *extstate.StateDB) error {
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/export.go:145]: From office discussion, the PRs may be out of order, but can we document somewhere (not necessarily here) that non AVAX ANTs are no longer supported?

File: vms/saevm/cchain/tx/tx.go (hunk 3 above)
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx.go:47]: maybe here is a godo place.

File: vms/saevm/cchain/tx/compatibility_test.go [hunk 6]
```diff
@@ -15,19 +16,29 @@ import (
...
+// fuzzer, and fuzzes the test.
 func fuzz(f *testing.F, ff func(t *testing.T, tx *Tx)) {
 	fuzzer := &txtest.F{
 		F: f,
+		Addresses: []common.Address{
+			{1},
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/compatibility_test.go:36]: Can we add another address so our seed corpus has multiple accounts?

File: vms/saevm/cchain/tx/compatibility_test.go [hunk 7]
```diff
@@ -165,3 +174,61 @@ func FuzzAtomicRequestsCompatibility(f *testing.F) {
...
+					state.AddBalanceMultiCoin(in.Address, common.Hash(in.AssetID), largeBigInt())
+				}
+			}
+		}
+
+		var (
+			oldTx = ToOldTx(t, newTx)
+			ctx   = &snow.Context{AVAXAssetID: AVAXAssetID}
+		)
+		require.NoError(t, oldTx.UnsignedAtomicTx.EVMStateTransfer(ctx, oldState))
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/compatibility_test.go]: (nit) Arran likes having these identify the call-site for readability in local testing I believe, just throwing it out there.

File: vms/saevm/cchain/tx/tx_test.go [hunk 8]
```diff
@@ -1310,3 +1322,182 @@ func TestAtomicRequests(t *testing.T) {
...
+func NewEmptyStateDB(t testing.TB) *extstate.StateDB {
+	t.Helper()
+
+	db := state.NewDatabase(rawdb.NewMemoryDatabase())
+	sdb, err := state.New(types.EmptyRootHash, db, nil)
+	require.NoError(t, err)
+	return extstate.New(sdb)
+}
+
+func TestTransferNonAVAX(t *testing.T) {
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx_test.go:1350]: @claude comment left to me, when asked about test coverage. 

> no Import test verifies that pre-existing balances are preserved

seems easy enough?

File: vms/saevm/cchain/tx/tx_test.go (hunk 2 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx_test.go]: I switched to just register all the libevm extras. I don't think we gain much useful context by just registering the stuff we care about here.

File: vms/saevm/cchain/tx/tx_test.go [hunk 9]
```diff
@@ -1310,3 +1322,182 @@ func TestAtomicRequests(t *testing.T) {
...
+		alice = common.Address{1}
+		bob   = common.Address{2}
+		btc   = ids.ID{3}
+		eth   = ids.ID{4}
+	)
+	tests := []struct {
+		name    string
+		init    map[common.Address]map[ids.ID]uint64
+		tx      Unsigned
+		want    map[common.Address]map[ids.ID]uint64
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/tx_test.go:1361]: Would it be possible to construct a `want` and then do a full comparison, rather than the balance fields one at a time?

**Review by JonathanOppenheimer — COMMENTED** [review]
I haven't read #5303, or the next in the series so some of this may be irrelevant.

File: vms/saevm/cchain/tx/tx.go (hunk 3 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/tx.go:47]: Just having one would be cool!

**Review by alarso16 — APPROVED** [review]

File: vms/saevm/cchain/tx/compatibility_test.go [hunk 10]
```diff
@@ -165,3 +174,61 @@ func FuzzAtomicRequestsCompatibility(f *testing.F) {
...
+		if diff := cmp.Diff(oldState, newState, opts...); diff != "" {
+			t.Errorf("%T.TransferNonAVAX() diff (-want +got):\n%s", newTx, diff)
+		}
+	})
+}
+
+// largeUint256 and largeBigInt return a balance large enough to never underflow
+// but small enough to never overflow during test arithmetic.
+func largeUint256() *uint256.Int { return new(uint256.Int).Lsh(uint256.NewInt(1), 128) }
+func largeBigInt() *big.Int      { return new(big.Int).Lsh(big.NewInt(1), 128) }
```
Comment (JonathanOppenheimer) [inline vms/saevm/cchain/tx/compatibility_test.go:222]: Also, if these are general use, may ever be re-used, would it make sense to move this to `vms/saevm/saetest`? I'm reminded of the zero/value package.

File: vms/saevm/cchain/tx/compatibility_test.go (hunk 4 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/compatibility_test.go:30]: So, I think this is a bug.. I'm not really sure what it's a bug in (gopls?) but linking should work for package variables... So I don't know why this doesn't link.

I suspect that the tool doesn't handle exported identifiers from test packages with dot imports... Pretty niche and specific... Perfect for a mishandled edge-case haha

File: vms/saevm/cchain/tx/export.go (hunk 5 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:145]: I'll add a comment on the interface.

File: vms/saevm/cchain/tx/compatibility_test.go (hunk 6 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/compatibility_test.go:36]: Talked offline, we already include multiple accounts during the fuzzing (just not necessarily multiple duplicate accounts)

File: vms/saevm/cchain/tx/compatibility_test.go (hunk 10 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/compatibility_test.go:222]: If we need these later we can consider it. I don't think it makes sense to move them out for a "maybe" in the future.

File: vms/saevm/cchain/tx/tx.go
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/tx.go:59]: I added a comment.

File: vms/saevm/cchain/tx/compatibility_test.go (hunk 4 above)
Comment (JonathanOppenheimer) [reply vms/saevm/cchain/tx/compatibility_test.go:30]: Ah unfortunate. I guess leave it as then and hopefully it will be eventually fixed in `gopls` or wherever it's broken?

File: vms/saevm/cchain/tx/export.go (hunk 1 above)
Comment (StephenButtolph) [reply vms/saevm/cchain/tx/export.go:16]: I don't think it would make sense to try to add some interface indirection. That really only works if you want to own the top and bottom of the stack, but not the middle... For SAE, we want to own the top (but not the bottom) of the stack... So we would need to import graft anyways in that world.

I agree that importing graft is something we need to do very sparingly, but I don't think adding a file with a bunch of aliases really helps us that much... It'll be very easy to see where graft is used (by just searching for the imports) and we will need to inspect the usage as we go around whether or not it makes sense to actually depend on that component within graft.

**Review by JonathanOppenheimer — APPROVED** [review]
